src/iter_arc.rs

```diff
@@ -0,0 +1,101 @@
+use triomphe::Arc;
+
+use crate::{
+    Leaf, Tree, Value,
+    utils::{Length, opt_packing_depth},
+};
+
+#[derive(Debug)]
+pub struct ArcIter<'a, T: Value> {
+    /// Stack of tree nodes corresponding to the current position.
+    stack: Vec<&'a Tree<T>>,
+    /// The list index corresponding to the current position (next element to be yielded).
+    index: usize,
+    /// The `depth` of the root tree.
+    full_depth: usize,
+    /// Cached packing depth to avoid re-calculating `opt_packing_depth`.
+    packing_depth: usize,
+    /// Number of items that will be yielded by the iterator.
+    length: Length,
+}
+
+impl<'a, T: Value> ArcIter<'a, T> {
+    pub fn from_index(index: usize, root: &'a Tree<T>, depth: usize, length: Length) -> Self {
```

Comment:
This should probably check `T::tree_hash_type` and return a `Result`. At the moment it's a bit weird that `Vector::iter_arc` can't fail.